runtime/containerd/containerd.go

```diff
@@ -0,0 +1,756 @@
+			"clab.ipv4.netmask": strconv.Itoa(ipv4nm),
+			"clab.ipv6.addr":    ipv6,
+			"clab.ipv6.netmask": strconv.Itoa(ipv6nm),
+		}
+		_, err = newContainer.SetLabels(ctx, additionalLabels)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func cniInit(cId, ifName string, mgmtNet *types.MgmtNet) (*libcni.CNIConfig, *libcni.NetworkConfigList, *libcni.RuntimeConf, error) {
+	// allow overwriting cni plugin binary path via ENV var
+
+	cnic := libcni.NewCNIConfigWithCacheDir([]string{cniPath}, cniCache, nil)
+
+	cniConfig := fmt.Sprintf(`
+	{
+		"cniVersion": "0.4.0",
+		"name": "clabmgmt",
+		"plugins": [
+		  {
+			"type": "bridge",
+			"bridge": "%s",
+			"isDefaultGateway": true,
+			"forceAddress": false,
+			"ipMasq": true,
+			"hairpinMode": true,
+			"ipam": {
+			  "type": "host-local",
+			  "ranges": [
+				[
+				  {
+					"subnet": "%s"
+				  }
+				],
+				[
+				  {
+					"subnet": "%s"
+				  }
+				]
+			  ]
+			}
+		  },
+		  {
+			"type": "tuning",
+			"mtu": %s,
+			"capabilities": {
+			  "mac": true
+			}
+		  },
+		  {
+			"type": "portmap",
+			"capabilities": {
+			  "portMappings": true
+			}
+		  }
+		]
+	  }
+	`, mgmtNet.Bridge, mgmtNet.IPv4Subnet, mgmtNet.IPv6Subnet, mgmtNet.MTU)
+	//log.Debug(cniConfig)
+	cncl, err := libcni.ConfListFromBytes([]byte(cniConfig))
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	cnirc := &libcni.RuntimeConf{
+		ContainerID: cId,
+		IfName:      ifName,
+		//// NetNS must be set later, can just be determined after cotnainer start
+		//NetNS:          node.NSPath,
+		CapabilityArgs: make(map[string]interface{}),
+	}
+	return cnic, cncl, cnirc, nil
+}
+
+type portMapping struct {
+	HostPort      int    `json:"hostPort"`
+	HostIP        string `json:"hostIP,omitempty"`
+	ContainerPort int    `json:"containerPort"`
+	Protocol      string `json:"protocol"`
+}
+
+func WithSysctls(sysctls map[string]string) oci.SpecOpts {
+	return func(ctx context.Context, client oci.Client, c *containers.Container, s *specs.Spec) error {
+		if s.Linux == nil {
+			s.Linux = &specs.Linux{}
+		}
+		if s.Linux.Sysctl == nil {
+			s.Linux.Sysctl = make(map[string]string)
+		}
+		for k, v := range sysctls {
+			s.Linux.Sysctl[k] = v
+		}
+		return nil
+	}
+}
+
+func (c *ContainerdRuntime) StartContainer(ctx context.Context, containername string) error {
+	container, err := c.client.LoadContainer(ctx, containername)
+	if err != nil {
+		return err
+	}
+	task, err := container.NewTask(ctx, cio.LogFile("/tmp/clab/"+containername+".log"))
+	if err != nil {
+		log.Fatal(err)
+		log.Fatalf("Failed to start container %s", containername)
+
+		return err
+	}
+	err = task.Start(ctx)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+func (c *ContainerdRuntime) StopContainer(ctx context.Context, containername string, dur *time.Duration) error {
+	ctask, err := c.getContainerTask(ctx, containername)
+	if err != nil {
+		log.Debugf("no task found for container %s", containername)
+		return nil
```

Comment:
return err ? and add it to the debug log statement ?

Reply:
well if the container is to be deleted but the task already died ... we log it and exit without error ... because the container is already stopped ...
So to me this looks fine.

Reply:
is the only possible error coming from the task dying ? or task not found?
the error itself it not logged

Reply:
using the following now:
```
	ctask, err := c.getContainerTask(ctx, containername)
	if err != nil {
		log.Debugf("container %s: %v", containername, err)
		return nil
	}
```